apollo-router/src/query_planner/query_planner_service.rs

```diff
@@ -184,7 +184,16 @@ impl QueryPlannerService {
             let result = result.map_err(FederationErrorBridge::from);
 
             let elapsed = start.elapsed().as_secs_f64();
-            metric_query_planning_plan_duration(RUST_QP_MODE, elapsed);
+            match &result {
+                Ok(_) => metric_query_planning_plan_duration(RUST_QP_MODE, elapsed, "success"),
+                Err(FederationErrorBridge::Cancellation(e)) if e.contains("timeout") => {
```

Comment:
I'm wondering if there's no better way to handle this error ? With a type instead of trying to check if the error message contains `timeout` ?

Reply:
Sadly, FederationErrorBridge is stringly-typed:

```rust
pub(crate) enum FederationErrorBridge {
    /// {0}
    UnknownOperation(String),
    /// {0}
    OperationNameNotProvided(String),
    /// {0}
    Other(String),
    /// {0}
    Cancellation(String),
}
```

Reply:
I could try adding a flag to it that gets set at creation if you like.

Reply:
We could add a new variant to `SingleFederationError` or update the existing cancelled variant but the function responsible for checking for cancellation isn't aware of the reason for cancellation and I'm not sure how to communicate it.

Reply:
Another option: I could remove the timeout heuristic for now and we could just say that it was cancelled.

Reply:
If there's no other viable option let's keep it like this